pkg/functions/runner.go

```diff
@@ -0,0 +1,221 @@
+package functions
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"time"
+)
+
+const (
+	defaultRunHost        = "127.0.0.1" // TODO allow to be altered via a runOpt
+	defaultRunPort        = "8080"
+	defaultRunDialTimeout = 2 * time.Second
+	defaultRunStopTimeout = 10 * time.Second
+	readinessEndpoint     = "/health/readiness"
+
+	// defaultRunTimeout is long to allow for slow-starting functions by default
+	// TODO: allow to be shortened as-needed using a runOption.
+	defaultRunTimeout = 5 * time.Minute
+)
+
+type defaultRunner struct {
+	client *Client
+	out    io.Writer
+	err    io.Writer
+}
+
+func newDefaultRunner(client *Client, out, err io.Writer) *defaultRunner {
+	return &defaultRunner{
+		client: client,
+		out:    out,
+		err:    err,
+	}
+}
+
+func (r *defaultRunner) Run(ctx context.Context, f Function) (job *Job, err error) {
+	var (
+		port    = choosePort(defaultRunHost, defaultRunPort, defaultRunDialTimeout)
+		runFn   func() error
+		verbose = r.client.verbose
+	)
+
+	// Job contains metadata and references for the running function.
+	job, err = NewJob(f, defaultRunHost, port, nil, nil, verbose)
+	if err != nil {
+		return
+	}
+
+	// Scaffold the function such that it can be run.
+	if err = r.client.Scaffold(ctx, f, job.Dir()); err != nil {
+		return
+	}
+
+	// Runner for the Function's runtime.
+	if runFn, err = runFunc(ctx, job); err != nil {
+		return
+	}
+
+	// Run the scaffolded function asynchronously.
+	if err = runFn(); err != nil {
+		return
+	}
+
+	// Wait for it to become available before returning the metadata.
+	err = waitFor(job, defaultRunTimeout)
+	return
+}
+
+// runFunc returns a function which will run the user's Function based on
+// the jobs runtime.
+func runFunc(ctx context.Context, job *Job) (runFn func() error, err error) {
+	runtime := job.Function.Runtime
+	switch runtime {
```

Comment:
Please add `"typescript"` as a case for this switch statement, returning `runnerNotImplemented`.